repl/src/main/scala/ammonite/repl/interp/Interpreter.scala

```diff
@@ -311,14 +349,66 @@ class Interpreter(prompt0: Ref[String],
       (classFiles, newImports) <- compiled
       _ = Timer("cachedCompileBlock 4")
       cls <- eval.loadClass(fullyQualifiedName, classFiles)
-    } yield (cls, newImports)
+    } yield (cls, newImports, tag)
+
   }
 
-  def processModule(source: ImportHook.Source, code: String, wrapperName: Name, pkgName: Seq[Name]) = {
 
-    processModule0(source, code, wrapperName, pkgName, predefImports)
+  def evalCachedClassFiles(cachedData: Seq[CompileCache], pkg: String, wrapper: String): Unit = {
+
+    def evalMain(cls: Class[_]) =
+      cls.getDeclaredMethod("$main").invoke(null)
+
+    var blockNumber = 1;
+    def getBlockNumber = blockNumber match {
+      case 1 => ""
+      case _ => "_" + blockNumber.toString
+    }
+
+    // Iterate through cached blocks and execute classFiles
+    // blockNumber keeps track of blockIndex
+    cachedData.foreach { d =>
+      for {
+        cls <- eval.loadClass(pkg + "." + wrapper + getBlockNumber, d._1)
+      } yield {
+        evalMain(cls)
+      }
+      blockNumber += 1
+    }
   }
 
+
+  def processModule(source: ImportHook.Source,
+                    code: String,
+                    wrapperName: Name,
+                    pkgName: Seq[Name]
+                   ): Res[Imports] = if(scriptCaching) {
+    val cacheTag = "cache" + Util.md5Hash(Iterator(code.getBytes)).map("%02x".format(_)).mkString
+    storage.asInstanceOf[Storage.Folder].classFilesListLoad(pkgName.map(_.backticked).mkString("."), wrapperName.backticked, cacheTag) match {
+      case Seq() =>
+        init()
+        val res = processModule0(source, code, wrapperName, pkgName, predefImports)
+       res match{
+         case Res.Success(data) =>
+           val Imports(imports) = data._1
+           val files = for (d <- data._2) yield (d._1, d._2)
+           if (imports.count(_.toName == "main") == 0)
+             storage.asInstanceOf[Storage.Folder].classFilesListSave(pkgName.map(_.backticked).mkString("."), wrapperName.backticked, files, cacheTag)
+           res.map(_._1)
+         case r => res.asInstanceOf[Res[Imports]]
+       }
+      case cachedData =>
+        evalCachedClassFiles(cachedData, pkgName.map(_.backticked).mkString("."), wrapperName.backticked)
```

Comment:
If `evalCachedClassFiles` returns a `Res[_]`, we should propagate that `Res` to be the result of this function